src/server/pfs/server/master.go

```diff
@@ -34,9 +42,122 @@ func (d *driver) master(ctx context.Context) {
 			gc := chunk.NewGC(d.storage.ChunkStorage())
 			return gc.RunForever(ctx)
 		})
+		eg.Go(func() error {
+			return d.finishCommits(ctx)
+		})
 		return eg.Wait()
 	}, backoff.NewInfiniteBackOff(), func(err error, _ time.Duration) error {
 		log.Errorf("error in pfs master: %v", err)
 		return nil
 	})
 }
+
+func (d *driver) finishCommits(ctx context.Context) error {
+	return d.commits.ReadOnly(ctx).WatchF(func(ev *watch.Event) error {
+		if ev.Type == watch.EventError {
+			return ev.Err
+		}
+		var key string
+		commitInfo := &pfs.CommitInfo{}
+		if err := ev.Unmarshal(&key, commitInfo); err != nil {
+			return err
+		}
+		if commitInfo.Finishing == nil || commitInfo.Finished != nil {
+			return nil
+		}
+		return backoff.RetryUntilCancel(ctx, func() error {
+			id, err := d.getFileSet(ctx, commitInfo.Commit)
+			if err != nil {
+				if pfsserver.IsCommitNotFoundErr(err) {
+					return nil
+				}
+				return err
+			}
+			// Compact the commit.
+			compactedId, err := d.compactor.Compact(ctx, []fileset.ID{*id}, defaultTTL)
+			if err != nil {
+				return err
+			}
+			if err := d.commitStore.SetTotalFileSet(ctx, commitInfo.Commit, *compactedId); err != nil {
+				return err
+			}
+			// Validate the commit.
+			// TODO(2.0 optional): Improve the performance of this by doing a logarithmic lookup per new file,
+			// rather than a linear scan through all of the files.
+			fs, err := d.storage.Open(ctx, []fileset.ID{*compactedId})
+			if err != nil {
+				return err
+			}
+			var prev string
+			var size int64
+			if err := fs.Iterate(ctx, func(f fileset.File) error {
+				idx := f.Index()
+				if prev != "" && (idx.Path == prev || strings.HasPrefix(idx.Path, prev+"/")) {
+					commitInfo.Error = true
+				}
+				prev = idx.Path
+				size += index.SizeBytes(idx)
+				return nil
+			}); err != nil {
+				return err
+			}
+			// Finish the commit.
+			return d.txnEnv.WithWriteContext(ctx, func(txnCtx *txncontext.TransactionContext) error {
+				commitInfo.Finished = txnCtx.Timestamp
+				commitInfo.SizeBytesUpperBound = size
+				if commitInfo.Details == nil {
+					commitInfo.Details = &pfs.CommitInfo_Details{}
+				}
+				commitInfo.Details.SizeBytes = size
+				if err := d.commits.ReadWrite(txnCtx.SqlTx).Put(pfsdb.CommitKey(commitInfo.Commit), commitInfo); err != nil {
```

Comment:
Do we have an update method anywhere, or do we always do a Get then Put?

Reply:
I will make this an update.